rust/src/storage/s3/dynamodb.rs

```diff
@@ -0,0 +1,427 @@
+use std::collections::HashMap;
+use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};
+
+use rusoto_core::RusotoError;
+use rusoto_dynamodb::*;
+use uuid::Uuid;
+
+pub struct DynamoLock {
+    client: DynamoDbClient,
+    opts: Options,
+}
+
+#[derive(Clone, Debug)]
+pub struct Options {
+    pub partition_key_value: String,
+    pub table_name: String,
+    pub owner_name: String,
+    pub lease_duration: u64,
+    pub sleep_millis: u64,
+}
+
+#[derive(Clone, Debug)]
+pub struct LockItem {
+    pub owner_name: String,
+    pub record_version_number: String,
+    pub lease_duration: u64,
+    pub is_released: bool,
+    pub data: Option<String>,
+    pub lookup_time: u128,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum DynamoError {
+    #[error("Dynamo table not found")]
+    TableNotFound,
+
+    #[error("Conditional check failed")]
+    ConditionalCheckFailed,
+
+    #[error("DynamoDB item has invalid schema")]
+    InvalidItemSchema,
+
+    #[error("Could not acquire lock for {0} sec")]
+    TimedOut(u64),
+
+    #[error("Put item error: {0}")]
+    PutItemError(RusotoError<PutItemError>),
+
+    #[error("Update item error: {0}")]
+    UpdateItemError(#[from] RusotoError<UpdateItemError>),
+
+    #[error("Get item error: {0}")]
+    GetItemError(RusotoError<GetItemError>),
+
+    #[error("Delete item error: {0}")]
+    DeleteItemError(#[from] RusotoError<DeleteItemError>),
+}
+
+pub const PARTITION_KEY_NAME: &str = "key";
+pub const OWNER_NAME: &str = "ownerName";
+pub const RECORD_VERSION_NUMBER: &str = "recordVersionNumber";
+pub const IS_RELEASED: &str = "isReleased";
+pub const LEASE_DURATION: &str = "leaseDuration";
+pub const DATA: &str = "data";
+
+mod expressions {
+    pub const ACQUIRE_LOCK_THAT_DOESNT_EXIST: &str = "attribute_not_exists(#pk)";
+
+    pub const PK_EXISTS_AND_IS_RELEASED: &str = "attribute_exists(#pk) AND #ir = :ir";
+
+    pub const PK_EXISTS_AND_RVN_MATCHES: &str = "attribute_exists(#pk) AND #rvn = :rvn";
+
+    pub const PK_EXISTS_AND_OWNER_RVN_MATCHES: &str =
+        "attribute_exists(#pk) AND #rvn = :rvn AND #on = :on";
+
+    pub const UPDATE_IS_RELEASED_AND_DATA: &str = "SET #ir = :ir, #d = :d";
+
+    pub const UPDATE_IS_RELEASED: &str = "SET #ir = :ir";
+}
+
+mod vars {
+    pub const PK_PATH: &str = "#pk";
+    pub const RVN_PATH: &str = "#rvn";
+    pub const RVN_VALUE: &str = ":rvn";
+    pub const IS_RELEASED_PATH: &str = "#ir";
+    pub const IS_RELEASED_VALUE: &str = ":ir";
+    pub const OWNER_NAME_PATH: &str = "#on";
+    pub const OWNER_NAME_VALUE: &str = ":on";
+    pub const DATA_PATH: &str = "#d";
+    pub const DATA_VALUE: &str = ":d";
+}
+
+mod options {
+    pub const PARTITION_KEY_VALUE: &str = "DYNAMO_LOCK_PARTITION_KEY_VALUE";
+    pub const TABLE_NAME: &str = "DYNAMO_LOCK_TABLE_NAME";
+    pub const OWNER_NAME: &str = "DYNAMO_LOCK_OWNER_NAME";
+    pub const LEASE_DURATION: &str = "DYNAMO_LOCK_LEASE_DURATION";
+    pub const SLEEP_MILLIS: &str = "DYNAMO_LOCK_SLEEP_MILLIS";
```

Comment:
could you add doc strings to all the public interfaces? this is something we will be enforcing after https://github.com/delta-io/delta-rs/pull/156 is merged.

Reply:
btw, is it possible to just copy paste some of the docs from java origin impl?

Reply:
yes, that's perfectly fine as long as they are applicable. we copy stuff from delta spec into our doc string as well :)